src/Renci.SshNet/PageantProtocol.cs

```diff
@@ -0,0 +1,200 @@
+﻿#if NETFRAMEWORK && !NET20 && !NET35
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Net;
+using System.Runtime.InteropServices;
+using System.Text;
+using Renci.SshNet.Common;
+
+namespace Renci.SshNet.Pageant {
+    /// <summary>
+    /// 
+    /// </summary>
+    public class PageantProtocol : IAgentProtocol {
+
+        #region  Constants
+
+        private const int WM_COPYDATA = 0x004A;
+
+        private const long AGENT_COPYDATA_ID = 0x804e50ba;
+        private const int AGENT_COPYDATA_ID_INT = unchecked ((int) AGENT_COPYDATA_ID);
+
+        private const int AGENT_MAX_MSGLEN = 8192;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public const byte SSH2_AGENTC_REQUEST_IDENTITIES = 11;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public const byte SSH2_AGENT_IDENTITIES_ANSWER = 12;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public const byte SSH2_AGENTC_SIGN_REQUEST = 13;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public const byte SSH2_AGENT_SIGN_RESPONSE = 14;
+
+        #endregion
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public static bool IsRunning {
+            get {
+                var hWnd = NativeMethods.FindWindow ("Pageant", "Pageant");
+
+                return hWnd != IntPtr.Zero;
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public PageantProtocol () {
+            var hWnd = NativeMethods.FindWindow ("Pageant", "Pageant");
+
+            if (hWnd == IntPtr.Zero) {
+                throw new SshException ("Pageant not running");
+            }
+
+        }
+
+        #region Implementation of IAgentProtocol
+
+        IEnumerable<IdentityReference> IAgentProtocol.GetIdentities () {
+            var hWnd = NativeMethods.FindWindow ("Pageant", "Pageant");
+
+            if (hWnd == IntPtr.Zero) {
+                yield break;
+            }
+
+            string mmFileName = Path.GetRandomFileName ();
+
+            using (var mmFile = MemoryMappedFile.CreateNew (mmFileName, AGENT_MAX_MSGLEN)) {
+                var security = mmFile.GetAccessControl ();
+                security.SetOwner (System.Security.Principal.WindowsIdentity.GetCurrent ().User);
+                mmFile.SetAccessControl (security);
+                using (var accessor = mmFile.CreateViewAccessor ()) {
+
+                    accessor.Write (0, IPAddress.NetworkToHostOrder (AGENT_MAX_MSGLEN - 4));
```

Comment:
Here would a be good place to have code comments to explain what we are doing, point to documentation and/or specification that apply to this implementation.